puffin/src/scope_details.rs

```diff
@@ -0,0 +1,261 @@
+        self.scope_id = Some(scope_id);
+        self
+    }
+
+    // This function should not be exposed as users are supposed to provide scope name in constructor.
+    #[inline]
+    pub(crate) fn with_scope_name<T>(mut self, scope_name: T) -> Self
+    where
+        T: Into<Cow<'static, str>>,
+    {
+        self.scope_name = Some(scope_name.into());
+        self
+    }
+}
+
+impl Serialize for ScopeDetails {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct(stringify!(ScopeDetails), 5)?;
+        state.serialize_field("scope_id", &self.scope_id)?;
+        state.serialize_field("scope_name", &self.scope_name)?;
+        state.serialize_field("function_name", &self.function_name)?;
+        state.serialize_field("file_path", &self.file_path)?;
+        state.serialize_field("line_nr", &self.line_nr)?;
+        state.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for ScopeDetails {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct ScopeDetailsOwnedVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for ScopeDetailsOwnedVisitor {
+            type Value = ScopeDetails;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str(&format!("struct {}", stringify!(ScopeDetails)))
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let scope_id = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::missing_field("scope_id"))?;
+                let scope_name = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::missing_field("scope_name"))?;
+                let function_name = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::missing_field("function_name"))?;
+                let file_path = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::missing_field("file_path"))?;
+                let line_nr = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::missing_field("line_nr"))?;
+
+                Ok(ScopeDetails {
+                    scope_id,
+                    scope_name,
+                    function_name,
+                    file_path,
+                    line_nr,
+                })
+            }
+        }
+
+        const FIELDS: &[&str] = &[
+            "scope_id",
+            "scope_name",
+            "function_name",
+            "file_path",
+            "line_nr",
+        ];
+        deserializer.deserialize_struct(stringify!(ScopeDetails), FIELDS, ScopeDetailsOwnedVisitor)
+    }
+}
```

Comment:
Again, how is this different from `#[derive(serde::Serialize)]` ?

Reply:
Cow doesnt imple serde so need to do a manual implementation?

Reply:
Ah before I had statics here and it was not working well with deserialize trait bu now I can just use the default impl